RPC plugin management

darosior opened the pull request:
~~This PR adds a new RPC command : `reloadplugins`, which will initialize and configure each plugin from the default directory (`~/.lightning/plugins/`) *-or from another passed as parameter-* that were not already loaded at lightningd startup.~~

This PR has mutated to now add a `plugin` RPC command : https://github.com/ElementsProject/lightning/pull/2771#issuecomment-505914017.

This permits to manage plugins without having to restart lightningd.

rustyrussell commented:
Can we have a top-level "plugin" command, with "stop", "start", "adddir"? Then we can stop and start them individually or all?

ZmnSCPxj commented on lightningd/plugin.c:
```diff
@@ -974,11 +986,11 @@ void plugins_init(struct plugins *plugins, const char *dev_plugin_debug)
 
-	plugins_add_default_dir(plugins, path_join(tmpctx, plugins->ld->config_dir, "plugins"));
-
-	setenv("LIGHTNINGD_PLUGIN", "1", 1);
```
Why is env removed?  In theory we do this so that a plugin could be run outside of `lightningd` (it could do something as simple as print "this is a lightningd plugin, not intended to be run directly as a command").

darosior replied:
This was not intended

ZmnSCPxj commented on lightningd/plugin.c:
```diff
@@ -1146,3 +1161,51 @@ struct log *plugin_get_log(struct plugin *plugin)
+	} else
+		plugins_add_default_dir(cmd->ld->plugins,
+					  path_join(tmpctx, cmd->ld->config_dir, "plugins"));
+	plugins_init(cmd->ld->plugins, cmd->ld->dev_debug_subprocess);
```
`plugin_init` clears the `pending_requests` field, which is used to route plugin responses back to the command (I think).  So if we execute `reload_plugin` while a plugin command is ongoing, this will prevent those command responses from being recognized one the plugin finishes the command, and also potentially leak the memory involved.

Should `reload_plugin` actively *kill* running plugins (and possibly raise an `error` on existing commands that are being handled by the plugin)?

darosior replied:
> plugin_init clears the pending_requests field, which is used to route plugin responses back to the command (I think). So if we execute reload_plugin while a plugin command is ongoing, this will prevent those command responses from being recognized once the plugin finishes the command, and also potentially leak the memory involved.

>

ZmnSCPxj replied:
> I thought that nobody would reload plugins while a command while pending..

Nobody might do so deliberately, but there exist such things as bugs.

darosior replied:
> I cannot see yet any code that kills plugins for reload_plugins, is this deliberate or did I miss?

Actually it only loads plugin which have not been loaded at launch. The name might not be accurate..

ZmnSCPxj replied:
`load_new_plugins`?

If so then running plugin-provided commands should be unaffected. Suggest refactoring the clearing of the fields away from starting new plugins then.

darosior commented:
> Can we have a top-level "plugin" command, with "stop", "start", "adddir"? Then we can stop and start them individually or all?
> 
> And list of course...

Yes, I'm not sure but I think I can do that.
@rustyrussell by the way, do you know why calling another function from [`lightningd/plugin.c`](https://github.com/darosior/lightning/blob/bf41958f90b2ff262d61ac2ebccba7c7a19974ab/lightningd/lightningd.c#L669) when having configured with `--enable-dev` causes a link error ?

@ZmnSCPxj thank you for the review

> You also need to consider what happens to ongoing plugin-based commands. Would not it be better if we enter a "shutting down" mode where existing plugin-provided commands are allowed to complete, new plugin-provided commands are delayed until after reload, and once current plugin commands have completed, only then kill and reload plugins?

I may add a check, even if it seems weird to me that someone would `reload_plugins` if a `withdraw` command is ongoing.

darosior commented:
According to the above comments by @rustyrussell and @ZmnSCPxj : 
- it is now a `plugin` command which takes a `command` and an optional `parameter`. The command returns in any case a list `active_plugins` of plugins loaded after the command execution.
    - The `start` command adds a plugin given a path as `parameter`
    - The `stop` command stops (kill and unload) a plugin given a plugin name as `parameter`
    - The `adddir` command adds a plugin directory given a directory path as `parameter`
    - The `reload` command adds all non-already-loaded plugins from the default plugins directory
    - The `list` command just output the active plugins
- it has been moved to a new file : `plugin_control.c`

darosior commented:
Rebased to move `plugins_add_directory()` back from `lightningd.c`'s `main()` to `plugins_init()` in `plugin.c` in order to make the `--enable-developer` build pass.

Since the load from the default plugin directory is now done each time we call a `plugin` command there is no more need for a dedicated `reload` command, so I removed it along with the above movement.

darosior commented:
The travis failure is due to `lightningd` crashing if the `plugin` command is used just after having started the daemon.
I tried (the above force pushs) to increase the `time.sleep(n)` in tests to make Travis tests pass (they pass locally) but I need to find the root cause of this crash in order to make a condition in the command so that `plugins_init()` is not called too early.
I've been investigating today and did not find it : the backtrace gets back to the `plugins_init()`'s `io_loop` to wait for manifests, but it appears that the loop is broken very early and that it's not the cause of the crash. Does anyone could explain me why we can't make another `io_loop` just after startup without waiting about 5 seconds ? Knowing that passed that delay, we can call the `plugin` command as many times in a row as we want ? Is disabling the command for 5s after startup (just to avoid the crash for the user) something that should be avoided ?

<details>
<summary> Backtrace to avoid pull and compilation </summary>
<p>

The backtrace after having quiet the warning notifications (`plugin_notify()`) which were part of the backtrace (memleak due to crash) without being the root cause.

```bash
2019-06-27T20:52:39.201Z lightningd(23821): --------------------------------------------------
2019-06-27T20:52:39.201Z lightningd(23821): Server started with public key 023bafa606486c3066f71d98f385dcfb24742191589922bcbea49373c21c4e7457, alias PEEVEDSOURCE-623-g726d6b1-modded (color #023baf) and lightningd v0.7.0-623-g726d6b1-modded
Already in transaction from lightningd/jsonrpc.c:653
2019-06-27T20:52:40.251Z lightningd(23821): Already in transaction from lightningd/jsonrpc.c:653
lightningd: FATAL SIGNAL 6 (version v0.7.0-623-g726d6b1-modded)
0x55af55bb0aa8 send_backtrace
	common/daemon.c:40
0x55af55bb0b4e crashdump
	common/daemon.c:53
0x7fa88ffdc83f ???
	???:0
0x7fa88ffdc7bb ???
	???:0
0x7fa88ffc7534 ???
	???:0
0x55af55b8be73 fatal
	lightningd/log.c:622
0x55af55bdd0ae db_begin_transaction_
	wallet/db.c:637
0x55af55ba990e sd_msg_read
	lightningd/subd.c:409
0x55af55bf64c9 next_plan
	ccan/ccan/io/io.c:59
0x55af55bf7046 do_plan
	ccan/ccan/io/io.c:407
0x55af55bf7084 io_ready
	ccan/ccan/io/io.c:417
0x55af55bf90d9 io_loop
	ccan/ccan/io/poll.c:445
0x55af55ba778d plugins_init
	lightningd/plugin.c:969
0x55af55ba8042 json_plugin_control
	lightningd/plugin_control.c:65
0x55af55b8729b parse_request
	lightningd/jsonrpc.c:654
0x55af55b876d5 read_json
	lightningd/jsonrpc.c:752
0x55af55bf64c9 next_plan
	ccan/ccan/io/io.c:59
0x55af55bf7046 do_plan
	ccan/ccan/io/io.c:407
0x55af55bf7084 io_ready
	ccan/ccan/io/io.c:417
0x55af55bf90d9 io_loop
	ccan/ccan/io/poll.c:445
0x55af55b83b0d io_loop_with_timers
	lightningd/io_loop_with_timers.c:24
0x55af55b89d43 main
	lightningd/lightningd.c:827
0x7fa88ffc909a ???
	???:0
0x55af55b720d9 ???
	???:0
0xffffffffffffffff ???
	???:0
2019-06-27T20:52:40.258Z lightningd(23821): FATAL SIGNAL 6 (version v0.7.0-623-g726d6b1-modded)
2019-06-27T20:52:40.258Z lightningd(23821): backtrace: common/daemon.c:45 (send_backtrace) 0x55af55bb0afe
2019-06-27T20:52:40.258Z lightningd(23821): backtrace: common/daemon.c:53 (crashdump) 0x55af55bb0b4e
2019-06-27T20:52:40.258Z lightningd(23821): backtrace: (null):0 ((null)) 0x7fa88ffdc83f
2019-06-27T20:52:40.258Z lightningd(23821): backtrace: (null):0 ((null)) 0x7fa88ffdc7bb
2019-06-27T20:52:40.258Z lightningd(23821): backtrace: (null):0 ((null)) 0x7fa88ffc7534
2019-06-27T20:52:40.258Z lightningd(23821): backtrace: lightningd/log.c:622 (fatal) 0x55af55b8be73
2019-06-27T20:52:40.258Z lightningd(23821): backtrace: wallet/db.c:637 (db_begin_transaction_) 0x55af55bdd0ae
2019-06-27T20:52:40.258Z lightningd(23821): backtrace: lightningd/subd.c:409 (sd_msg_read) 0x55af55ba990e
2019-06-27T20:52:40.258Z lightningd(23821): backtrace: ccan/ccan/io/io.c:59 (next_plan) 0x55af55bf64c9
2019-06-27T20:52:40.258Z lightningd(23821): backtrace: ccan/ccan/io/io.c:407 (do_plan) 0x55af55bf7046
2019-06-27T20:52:40.258Z lightningd(23821): backtrace: ccan/ccan/io/io.c:417 (io_ready) 0x55af55bf7084
2019-06-27T20:52:40.258Z lightningd(23821): backtrace: ccan/ccan/io/poll.c:445 (io_loop) 0x55af55bf90d9
2019-06-27T20:52:40.258Z lightningd(23821): backtrace: lightningd/plugin.c:969 (plugins_init) 0x55af55ba778d
2019-06-27T20:52:40.258Z lightningd(23821): backtrace: lightningd/plugin_control.c:65 (json_plugin_control) 0x55af55ba8042
2019-06-27T20:52:40.258Z lightningd(23821): backtrace: lightningd/jsonrpc.c:654 (parse_request) 0x55af55b8729b
2019-06-27T20:52:40.258Z lightningd(23821): backtrace: lightningd/jsonrpc.c:752 (read_json) 0x55af55b876d5
2019-06-27T20:52:40.258Z lightningd(23821): backtrace: ccan/ccan/io/io.c:59 (next_plan) 0x55af55bf64c9
2019-06-27T20:52:40.258Z lightningd(23821): backtrace: ccan/ccan/io/io.c:407 (do_plan) 0x55af55bf7046
2019-06-27T20:52:40.258Z lightningd(23821): backtrace: ccan/ccan/io/io.c:417 (io_ready) 0x55af55bf7084
2019-06-27T20:52:40.258Z lightningd(23821): backtrace: ccan/ccan/io/poll.c:445 (io_loop) 0x55af55bf90d9
2019-06-27T20:52:40.258Z lightningd(23821): backtrace: lightningd/io_loop_with_timers.c:24 (io_loop_with_timers) 0x55af55b83b0d
2019-06-27T20:52:40.258Z lightningd(23821): backtrace: lightningd/lightningd.c:827 (main) 0x55af55b89d43
2019-06-27T20:52:40.258Z lightningd(23821): backtrace: (null):0 ((null)) 0x7fa88ffc909a
2019-06-27T20:52:40.258Z lightningd(23821): backtrace: (null):0 ((null)) 0x55af55b720d9
2019-06-27T20:52:40.258Z lightningd(23821): backtrace: (null):0 ((null)) 0xffffffffffffffff
Log dumped in crash.log.20190627205240
Aborted
```

</p>

ZmnSCPxj commented:
Like I suggested, you should split `plugins_init`.  Initiating the startup of new plugins should be a separate function, which `plugins_init` calls, e.g. `plugins_start_plugins`.

In particular, you should not do an `io_loop` within an `io_loop`.

So `plugins_start_plugins` would be an async interface which calls the callback only when `pending_manifests` drops to 0.  Both `plugins_init` and `json_plugin_control` should call into it.  And `json_plugin_control` ***MUST NOT*** call `plugins_init`.

For `plugins_init`, it provides a callback which executes `io_break`.  `plugins_init` will also include its own `io_loop`.

For `json_plugin_control`, it provides a callback which provides the actual result object of the `plugin` command.  In particular it would ***not*** have its own `io_loop`.

darosior commented:
@ZmnSCPxj thank you the support : I had to modify `plugin.c` some more but it now works well.

I added the `reload` control back since `json_plugin_control` does not call `plugin_init` anymore. I also added a `unregister_hook` control that might be useful.

darosior commented:
@ZmnSCPxj do the changes seem good to you ?

ZmnSCPxj commented on lightningd/plugin_control.c:
```diff
@@ -0,0 +1,127 @@
+	"	loads not-already-loaded plugins from the default plugins dir\n"
+	"cli plugin list\n"
+	"	lists all active plugins\n"
+	"\n"
```
`unregister_hook` is not exist here?

darosior replied:
Forgot to add it to the verbose help.

ZmnSCPxj commented on lightningd/plugin_control.h:
```diff
@@ -0,0 +1,14 @@
+/**
+ * A plugin command which permits to control (start, stop, ..) plugins
+ */
+struct command_result *json_plugin_control(struct command *cmd,
```
Not need to make this public; can make `static` in the `plugin_control.c` file.

darosior replied:
Ok, is an emty header file OK ?

ZmnSCPxj replied:
Should be okay.

ZmnSCPxj commented on doc/lightning-plugin.7.txt:
```diff
@@ -0,0 +1,45 @@
+action to take and optionally one or two parameters which describes the plugin on which
+the action has to be taken.
+
+The 'start' command takes a path as parameter and will load the plugin available from this
```
The help description contains example, but manpage does not?

darosior replied:
I made a `cli command parameter` in verbose help to make it concise. I can replicate it in the manpage too, but it seems already more detailed.

ZmnSCPxj replied:
Generally manpage is more detailed and includes things like possible errors.  Maybe okay for now.

ZmnSCPxj commented on tests/test_plugin.py:
```diff
@@ -88,6 +88,48 @@ def test_plugin_dir(node_factory):
 
 
+def test_plugin_command(node_factory):
+    """Tests the 'reloadplugins' RPC command"""
```
Inaccurate comment.

darosior commented:
Thank you for the review @ZmnSCPxj. I added the forgotten `unregister_hook` verbose help, corrected the test comment and made `json_plugin_control()` static.

rustyrussell commented on contrib/pylightning/lightning/plugin.py:
```diff
@@ -108,7 +108,7 @@ def __init__(self, stdout=None, stdin=None, autopatch=True):
             self.stdin = sys.stdin
 
-        if os.getenv('LIGHTNINGD_PLUGIN') and autopatch:
+        if autopatch:
```
This patch is wrong, AFAICT.

darosior replied:
Since the `LIGHTNINGD_PLUGIN` was moved after the `io_loop` for startup plugins initialization, this would prevent some plugins to be initiated at startup. I'll amend this anyway.

rustyrussell commented on lightningd/plugin.c:
```diff
@@ -781,9 +792,9 @@ static void plugin_manifest_cb(const char *buffer,
+	 * if they have and this is the startup init */
 	plugin->plugins->pending_manifests--;
-	if (plugin->plugins->pending_manifests == 0)
+	if (plugin->plugins->pending_manifests == 0 && !getenv("LIGHTNINGD_PLUGIN"))
```
Abusing the env var to see if we're starting up is pretty horrible!

Better to have a 'startup' bool, or simply set pending_manifests to (size_t)-1 every time you start a plugin after startup, so this never happens?

darosior replied:
> Abusing the env var to see if we're starting up is pretty horrible!
Better to have a 'startup' bool, or simply set pending_manifests to (size_t)-1 every time you start a plugin after startup, so this never happens?

I did that because it was already used and to avoid yet another member, I'll amend it if it's preferable.

rustyrussell commented on lightningd/plugin.c:
```diff
@@ -943,17 +951,31 @@ void plugins_init(struct plugins *plugins, const char *dev_plugin_debug)
 	if (plugins->pending_manifests > 0)
 		io_loop_with_timers(plugins->ld);
+
+	setenv("LIGHTNINGD_PLUGIN", "1", 1);
```
You need to move this line back to before the io_loop, BTW.

darosior replied:
This was because of the above, to stop `io_break`ing after this loop.

rustyrussell commented on lightningd/plugin_control.c:
```diff
@@ -0,0 +1,135 @@
+		   p_req("command", param_string, &control_command),
+		   p_opt("parameter", param_string, &control_param),
+		   p_opt("second_parameter", param_string, &control_second_param),
+		   NULL))
```
We probably need more powerful param handling for cases like this, TBH.  This will do for now, however.

rustyrussell commented on lightningd/plugin_control.c:
```diff
@@ -0,0 +1,135 @@
+	if (streq(control_command, "stop")) {
+		plugin_found = false;
+		list_for_each(&cmd->ld->plugins->plugins, p, list) {
+			if (paths_match(p->cmd, control_param)) {
```
Rather than export paths_match, we should export 'find_plugin' which does this loop for us?

darosior replied:
`find_plugin` would abort on find failure IIRC

rustyrussell commented on lightningd/plugin_control.c:
```diff
@@ -0,0 +1,135 @@
+			return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+					   "Could not find plugin %s", control_param);
+	}
+	else if (streq(control_command, "start")) {
```
Nit: `else if` belongs on same line as `}`

darosior replied:
Because of the `else if` repetitions I took the liberty to add some space for readability. Will amend.

rustyrussell commented on lightningd/plugin_control.c:
```diff
@@ -0,0 +1,135 @@
+		else
+			return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+						   "Could not access %s with execution permission",
+						   control_param);
```
How about: `"%s is not executable: %s", control_param, strerror(errno))` which is more informative if the problem is something weird?

rustyrussell commented on lightningd/plugin_control.c:
```diff
@@ -0,0 +1,135 @@
+	else {
+		return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+					"Invalid control : %s should be one of"
+					" stop, start, adddir, reload",
```
... or unregister_hook?

rustyrussell commented on lightningd/plugin_control.c:
```diff
@@ -0,0 +1,135 @@
+		if (p->configured)
+			json_add_string(response, NULL, p->cmd);
+	}
+	json_array_end(response);
```
Might you want inactive plugins too?  How about a simple "plugins" with a 'active' boolean?

darosior replied:
Done

rustyrussell commented on lightningd/plugin_control.c:
```diff
@@ -0,0 +1,135 @@
+	json_plugin_control,
+	"Control plugins (start, stop, adddir, reload, list)",
+	.verbose = "Usage :\n"
+	"cli plugin start /path/to/a/plugin\n"
```
skip the 'cli' here, since it doesn't apply for non-lightning-cli users?

rustyrussell commented on lightningd/plugin_control.c:
```diff
@@ -0,0 +1,135 @@
+	"cli plugin stop plugin_name\n"
+	"	stops an already registered plugin\n"
+	"cli plugin adddir /path/to/a/plugin_dir/\n"
+	"	adds a new plugin directory\n"
```
`start` vs `adddir` is weird.  `add` vs `adddir` or `start` vs `startdir`?

rustyrussell replied:
Or maybe just `start` can take a filename or a directory?

darosior replied:
I chose `start` and `startdir` to avoid `ddd` which is error prone :)

rustyrussell commented on lightningd/plugin_control.c:
```diff
@@ -0,0 +1,135 @@
+	"cli plugin adddir /path/to/a/plugin_dir/\n"
+	"	adds a new plugin directory\n"
+	"cli plugin reload\n"
+	"	loads not-already-loaded plugins from the default plugins dir\n"
```
That is not what I expected to happen when I say `reload`; I expected it to take a plugin name to restart, or maybe `all` or nothing for everything?

darosior replied:
A plugin could be reloaded by `stop`ing and `start`ing it again, however I think it's a good design to involve the `plugins` directory and to make it possible to load all new plugins from this directory. Maybe the command name is not well chosen (as @ZmnSCPxj already pointed out)

rustyrussell replied:
"rescandir"?

darosior commented:
Thank you for the review @rustyrussell 
I rebased to : 
- remove the use of the (already present :) `LIGHTNINGD_PLUGIN` environment variable in favor of a `plugins` struct member to check for startup.
- add the inactive plugins to the command return. It knows output an object with a plugin name as key and its state (active/inactive) as value.
- some style/log/help nits.

rustyrussell commented:
With respect to @cdecker's concerns, I only somewhat agree.  Some plugins are fine with stop, start and reload, others are not.  I think it has to be an opt-in mechanism at getmanifest time: a new `"dynamic": true` field in the reply?

In addition, we should probably tell plugins, with `"startup": true` in the init call, that it's being called as part of the initial startup or not.  Separately, we might add a shutdown phase for plugins, but thats a different PR IMHO.

darosior commented:
> I'm not really convinced we want this at all:
> 
> * One of the guarantees that we currently have is that a plugin cannot miss a notification or hook, because they are configured and initialized before any can trigger. If we allow plugins to be started after `lightnignd` has been running for a while we lose that guarantee.

Is this guarantee required ? If one starts a plugin after `lightningd` has been running for a while he expects that the plugin won't receive notifications until it is fully initialized. 

> * Adding plugins really is a configuration option, not a runtime operation, so restarting the daemon is the correct way of adding a plugin in my mind.

I think this is powerful to be able to start plugins "on-the-fly" : some users might don't want to restart `lightningd` every time they want to add or remove a plugin.

> This is important since some plugins may rely on them being present after a restart (if we are managing invoices externally, or anything that uses a `hook` really). If we allow runtime loading of plugins we are breaking with that assumption.

This was why this PR initially permitted to only reload plugins from the plugins directory. If using the default plugins directory (or the `lightning/plugins` directory) to store plugins becomes standard then plugins added while `lightningd` was running will be loaded at the next startup. 

> * Being able to unregister a hook may also break a load of things: image having a backup plugin using the `db_write`, that gets unregistered and then re-registered after a while. It'd result in an incomplete backup that could be dangerous. Plugins registering hooks really expect the registration to be sticky. Let's rather add hook-result-combinators like @ZmnSCPxj proposed in #2796

I can remove this one. I understand that it potentially permits to break a load of things.
 
> That being said, you might also achieve the same goals by using the [autoreload](https://github.com/lightningd/plugins/tree/master/autoreload) plugin.

The [autoreload](https://github.com/lightningd/plugins/tree/master/autoreload) plugin seems to not allow to start new plugins after `lightningd` startup.




> With respect to @cdecker's concerns, I only somewhat agree. Some plugins are fine with stop, start and reload, others are not. I think it has to be an opt-in mechanism at getmanifest time: a new `"dynamic": true` field in the reply?
> 
> In addition, we should probably tell plugins, with `"startup": true` in the init call, that it's being called as part of the initial startup or not. Separately, we might add a shutdown phase for plugins, but thats a different PR IMHO.

I'll add this two fields, along with modifying the params to use #2814 if it worths that I continue updating this.

darosior commented:
Rebased, changed params to use #2814's subcommands, and removed the `unregister_hook` subcommand.

darosior commented:
Added the `getmanifest` and `init` fields as @rustyrussell [proposed](https://github.com/ElementsProject/lightning/pull/2771#issuecomment-511612602).

darosior commented:
I added a flag as adding a new field to `init` would break Python plugins using previous pylightning versions.

rustyrussell commented on contrib/pylightning/lightning/lightning.py:
```diff
@@ -714,6 +714,19 @@ def ping(self, peer_id, length=128, pongbytes=128):
+            payload["plugin"] = parameter
+        elif subcommand == "startdir":
+            payload["directory"] = parameter
+        return self.call("plugin", payload)
```
Hmm, discarding parameters if we don't know the subcommand seems wrong?  I think we need to whitelist the known parameters?

darosior replied:
I don't understand, do you mean whitelist on the C side ? Like accepting parameters for `reload` (soon `rescandir` :) and `list` ?
https://github.com/darosior/lightning/blob/09a2dea65292e6713aadeb60e253da8d0bd4f92c/lightningd/plugin_control.c#L81-L95

cdecker replied:
Maybe we should rewrite this into multiple calls that map to the various `plugin` subcommands?

 - plugin_start
 - plugin_stop
 - plugin_rescan
 - etc

That way we have something to attach the specific help messages to instead of handling opaque sets of parameters?

rustyrussell commented on lightningd/plugin.c:
```diff
@@ -1055,6 +1022,9 @@ static void plugin_config(struct plugin *plugin)
 	json_object_end(req->stream);
 
+	if (plugin->signal_startup)
+		json_add_bool(req->stream, "startup", plugin->plugins->startup);
```
I don't understand why having this set unconditionally breaks older pylightning?  We should be able to add init fields without repercussions; if we can't, can we fix that now so this won't be a problem in future?

darosior replied:
This is not a problem with plugins in general, but with how Pylightning (handfully) unpack fields into function. To avoid that now and in the future, we could either wrap the init response in an `init` object as in #2823 bit this would break all existing plugins.

cdecker replied:
Yeah, that argument unpacking was a bit of a brainfart on my end, but it has been working nicely so far. We should just define a minimal set of top-level objects that need to be passed in and then add new stuff to those. That also makes this semantically clean.

darosior replied:
Ok, i'll add `startup`to `configuration` then

rustyrussell commented on lightningd/plugin_control.c:
```diff
@@ -0,0 +1,132 @@
+{
+	const char *subcmd;
+	subcmd = param_subcommand(cmd, buffer, params,
+							  "start", "stop", "startdir", "reload", "list", NULL);
```
Weird indent here: do you have tabs set to 4 spaces?

darosior replied:
No, but when aligning I'm tricked by tabs

rustyrussell commented on lightningd/plugin_control.c:
```diff
@@ -0,0 +1,132 @@
+	json_object_start(response, "plugins");
+	list_for_each(&cmd->ld->plugins->plugins, p, list)
+		json_add_string(response, p->cmd, p->configured ? "active" : "inactive");
+	json_object_end(response);
```
We can't add fields to this, so it needs to be object-wrapped, even if we don't have other fields today, so it needs to be:
```
response = json_stream_success(cmd);
json_array_start(response, "plugins");
list_for_each(&cmd->ld->plugins->plugins, p, list) {
        json_object_start(response, NULL);
        json_add_bool(response, "active", p->configured);
        json_object_end(response);
}
```

In particular, in future we'll want to know what ones are dynamic, and in future what hooks they are registered for...

rustyrussell commented on contrib/pylightning/lightning/plugin.py:
```diff
@@ -99,7 +99,7 @@ class Plugin(object):
     """
 
-    def __init__(self, stdout=None, stdin=None, autopatch=True):
+    def __init__(self, stdout=None, stdin=None, autopatch=True, dynamic=True):
```
dynamic needs to default to False, I think.  Though perhaps there are only a few plugins which can simply be fixed.  @cdecker ?

darosior replied:
All plugins from the repo seems ok (even with `dynamic = True` they are not restarted.

cdecker replied:
Should be ok for now, I'll fixup the various plugins in `lightningd/plugins` once this gets merged.

darosior commented:
> and a question about adding init fields?

*(answering here too)*The problem is not with adding fields in the init response object but with how Pylightning conveniently unpack it to function parameters (cf #2823) : it has the side effect that a change in the object received will cause a break of the plugin if the callback Python function isn't amended to add a new parameter. We could wrap it into a top object as I propose for hooks and notifications in #2826 but this is a backward incompatible change.
However I think that's outside of the scope of this PR : we could handle this along with plugin shutdown (I have questions about this too but i'll make an issue to discuss this).

> 
> Oh, and this definitely deserves a CHANGELOG line :)

I'll add it :-)

darosior commented:
Rebased to : 
- correct indentation
- rename `reload` subcommand to `rescan`
- return an array of objects (`{ "name": "my_plugin", "active": true }`)

cdecker commented on lightningd/plugin.h:
```diff
@@ -1,28 +1,88 @@
+/**
+ * A plugin, exposed as a stub so we can pass it as an argument.
+ */
+struct plugin {
```
Why is it necessary to make all these things public? Wouldn't it be sufficient to make `plugin_kill` public?

darosior replied:
I think I need it for https://github.com/darosior/lightning/blob/99b4686406d5c79761cd03fe111b1140016afbe7/lightningd/plugin_control.c#L26

cdecker commented on lightningd/plugin.c:
```diff
@@ -813,6 +813,12 @@ static void plugin_manifest_cb(const char *buffer,
+	/* If all plugins have replied to getmanifest and this is not
+	 * the startup init, configure them */
+	if (!plugin->plugins->startup && plugin->plugins->pending_manifests == 0)
+		plugins_config(plugin->plugins);
```
Doesn't this result in us re-configuring all plugins every time we add a new plugin?

darosior replied:
No, 2ba5145a3e460684e5f65e5a09523238cf65dee5 introduces 
```C
list_for_each(&plugins->plugins, p, list) {
	if (p->configured)
		continue;
```

cdecker commented on doc/lightning-plugin.7.txt:
```diff
@@ -0,0 +1,48 @@
+-----------
+
+The *plugin* RPC command allows to manage plugins without having to restart lightningd.
+It takes 1 to 3 parameters: a command (start/stop/startdir/reload/list) which describes the
```
This still calls it `reload` whereas the code below calls it `rescan`

darosior replied:
Didnt see this one, thanks !

cdecker commented on .travis/build.sh:
```diff
@@ -10,6 +10,7 @@ export COMPAT=${COMPAT:-1}
 export PATH=$CWD/dependencies/bin:"$HOME"/.local/bin:"$PATH"
 export TIMEOUT=180
 export PYTEST_PAR=2
+export PYTHONPATH=$PWD/contrib/pylightning:$PYTHONPATH
```
This has always been working, what is the error you are getting without this?

darosior replied:
Dynamically-started plugins don't get the non-exported environment variable passed at the command line

cdecker commented on lightningd/plugin.c:
```diff
@@ -1014,6 +1019,8 @@ static void plugin_config(struct plugin *plugin)
 	json_add_string(req->stream, "rpc-file", ld->rpc_filename);
 	json_object_end(req->stream);
 
+	json_add_bool(req->stream, "startup", plugin->plugins->startup);
```
Adding startup like this will make all existing plugins incompatible since pylightning unpacks the arguments, and adding a new one will cause it to get an unknown argument. Can we make this part of the `configuration`?

darosior replied:
Yes sure, anyway I made a fixup afterwards :-). I can remove the fixup commit and make this field part of the `configuration`.

darosior commented:
> Yeah, that argument unpacking was a bit of a brainfart on my end, but it has been working nicely so far. We should just define a minimal set of top-level objects that need to be passed in and then add new stuff to those. That also makes this semantically clean.

Should we add this to `init` too ?

darosior commented:
Thank you for the review @cdecker , I rebased to (mainly) :
- Break Pylightning `pluging` method into one method by subcommand (which is much nicer ! :)
- Move the `init`'s `startup` field to its `configuration` object and thus remove the now useless fixup commit.

SimonVrouwe commented on contrib/pylightning/lightning/plugin.py:
```diff
@@ -99,7 +99,7 @@ class Plugin(object):
     """
 
-    def __init__(self, stdout=None, stdin=None, autopatch=True):
+    def __init__(self, stdout=None, stdin=None, autopatch=True, dynamic=True):
```
This was commented about earlier, but is it not more conservative to default to `dynamic=False` ?

Or is it now expected that old plugins need to update anyway to this new behavior?
Most of the plugins don't even have tests!

darosior replied:
> This was commented about earlier, but is it not more conservative to default to `dynamic=False` ?
> 
> Or is it now expected that old plugins need to update anyway to this new behavior?
> Most of the plugins don't even have tests!

This was discussed above. This does not change the plugin behavior at all, it just makes it possible for lightningd to restart it : this is completely transparent plugin side (hence the new init field).

darosior replied:
That said, a plugin that is not intended to be stopped at runtime can set dynamic to `False` to ensure that an user won't *inadvertently* stop it via RPC.

SimonVrouwe replied:
> This does not change the plugin behavior at all

Sorry to be blunt, but if the behavior is not changed, then why is there an extra parameter?
What is wrong with defaulting to `dynamic=False` ?

Maybe to you its trivial, but how can we know for sure that current (known or unknown) plugins work properly when re-starting with a running c-lightning?

darosior replied:
> Sorry to be blunt, but if the behavior is not changed, then why is there an extra parameter?
What is wrong with defaulting to dynamic=False ?

This is a new field passed in the `getmanifest` call, [used to give informations to lightningd about plugins](https://lightning.readthedocs.io/PLUGINS.html#the-getmanifest-method). This allows a plugin creator to pass whatever he wants.

> What is wrong with defaulting to dynamic=False ?

Pylightning users not setting it.

> Maybe to you its trivial, but how can we know for sure that current (known or unknown) plugins work properly when re-starting with a running c-lightning?

 Restarting a plugin means stopping and starting it : on the `stop` command, the plugin will be killed. On the `start` command an interface similar (but asynchronous) to which is setup on startup is setup by `lightningd` : a plugin must behave properly at `start` or it means it wouldn't even at startup.
The only case this could lead to an issue is a plugin which has to be loaded (for whatever reason, i.e. backup the first db calls) at startup *only* : that was `myplugin.dynamic = False` is for :-)

SimonVrouwe replied:
Ok thanks for the clarification, now I also read @cdecker's comment:

> Should be ok for now, I'll fixup the various plugins in lightningd/plugins once this gets merged.

darosior commented:
Rebased

rustyrussell commented:
> The only really important thing is the new arguments for pylightning which I need to find a fix for.

Would rpc.plugin.start() etc be more pythony?  ie. make plugin like a command namespace... somehow?

darosior commented:
> > The only really important thing is the new arguments for pylightning which I need to find a fix for.
> 
> Would rpc.plugin.start() etc be more pythony? ie. make plugin like a command namespace... somehow?

I think this was about dynamic json fields unpacking to function arguments which breaks plugins made with Pylightning if we add a new field (in `init` for example :-))

darosior commented:
Rebased on top of master and resolve CHANGELOG conflicts. Added new brag lines for `init` and `getmanifest` new fields.

rustyrussell commented on CHANGELOG.md:
```diff
@@ -12,6 +12,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - plugins: a new notification type `invoice_payment` (sent when an invoice is paid) has been added
+- rpc: a new rpc command is added, `plugin`. It allows one to manage plugins without restarting `lightningd`.
+- plugins: a new boolean field is added to the `init`'s `configuration`, `startup`. It allows a plugin to know if it has been started on `lightningd` startup.
+- plugins: a new boolean field can be added to a plugin manifest, `dynamic`. It allows a plugin to tell if it can be started or stopped "on-the-fly".
```
I'll probably collapse these three lines into "plugins: dynamic plugin support", which is the better summary for this level IMHO.

rustyrussell commented:
Trivial rebase to resolve CHANGELOG.md conflict...

SimonVrouwe commented:
Can someone add the `plugin` label to this threat?
